### PR TITLE
Reimplement light attenuation for spot and point lights

### PR DESCRIPTION
**Pull request by zzmp**
Reimplements light attenuation to simulate a point source.

Light attenuation had been implemented with an intensity threshold, such that a light surface radius was established to maintain that intensity. Because of this, light had a consistent brightness and then immediate dropoff.

Now, light attenuation is implemented as if from a source light, with an "apparent surface radius" governing the rate of attenuation (see model/Light.cpp for details). All spot/point lights will now appear as if from a point light.
The attenuation equation is biased by the "apparent surface radius" such that all light attenuates from the point, but the radius governs the rate.
There is still a minimum intensity cutoff. Its radius is calculated, after which the shaders discard the point/spot light contribution.

The new "apparent surface radius" is also exposed to the user in edit.js, through entityProperties.html.

Light now falls off from the source by 1/r^2. ***This looks drastically different.***
***This makes lots of things that had looked lit, appear unlit, as the light now drops off with 1/r^2.***

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 399861087d7cfca38df29bd0cadcb4fa76cd99d9 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

File: libraries/model/src/model/Light.cpp [hunk 1]
```diff
@@ -15,22 +15,26 @@ using namespace model;
 Light::Light() :
     _flags(0),
     _schemaBuffer(),
-    _transform() {
+    _transform(),
+    _maximumRadius(1.0f) {
```
Comment (jherico) [inline libraries/model/src/model/Light.cpp]: Coding standard 3.2.2, initializers should be put in the header, hence `float _maximumRadius { 1.0f };`.

File: libraries/model/src/model/Light.cpp [hunk 2]
```diff
@@ -15,22 +15,26 @@ using namespace model;
...
     _schemaBuffer = std::make_shared<gpu::Buffer>(sizeof(Schema), (const gpu::Byte*) &schema);
+    updateLightRadius();
 }
 
 Light::Light(const Light& light) :
     _flags(light._flags),
     _schemaBuffer(light._schemaBuffer),
-    _transform(light._transform) {
+    _transform(light._transform),
+    _maximumRadius(1.0f) {
```
Comment (jherico) [inline libraries/model/src/model/Light.cpp]: shouldn't this be `light._maximumRadius`?

File: libraries/model/src/model/Light.slh [hunk 3]
```diff
@@ -29,15 +29,7 @@ vec3 getLightColor(Light l) { return l._color.rgb; }
...
-float evalLightAttenuation(Light l, float r) {
-    float d = max(r - l._attenuation.x, 0.0);
-    float denom = d * l._attenuation.y + 1.0;
-    float attenuation = 1.0 / (denom * denom);
-    return max((attenuation - l._attenuation.z)/(1.0 - l._attenuation.z), 0.0);
-  //  return clamp(1.0/(l._attenuation.x + l._attenuation.y * r + l._attenuation.z * r * r), 0.0, 1.0);
-}
-
-float getLightSpotAngleCos(Light l) {
+ float getLightSpotAngleCos(Light l) {
```
Comment (jherico) [inline libraries/model/src/model/Light.slh]: extra whitespace

File: libraries/model/src/model/Light.slh [hunk 4]
```diff
@@ -49,15 +41,19 @@ float evalLightSpotAttenuation(Light l, float cosA) {
     return pow(cosA, l._spot.w);
 }
 
+float getLightRadius(Light l) {
+    return l._attenuation.x;
+}
+
 float getLightSquareRadius(Light l) {
```
Comment (jherico) [inline libraries/model/src/model/Light.slh]: `const Light&`

File: libraries/model/src/model/Light.slh [hunk 5]
```diff
@@ -49,15 +41,19 @@ float evalLightSpotAttenuation(Light l, float cosA) {
     return pow(cosA, l._spot.w);
 }
 
+float getLightRadius(Light l) {
```
Comment (jherico) [inline libraries/model/src/model/Light.slh]: `const Light& l`

File: libraries/model/src/model/Light.slh [hunk 6]
```diff
@@ -65,6 +61,16 @@ float getLightShowContour(Light l) {
     return l._control.w;
 }
 
+float evalLightAttenuation(Light l, float r) {
```
Comment (jherico) [inline libraries/model/src/model/Light.slh]: `const Light& l`

File: libraries/model/src/model/Light.slh (hunk 3 above)
Comment (jherico) [inline libraries/model/src/model/Light.slh]: extra whtespace

**Comment by hifi-gustavo** [comment]
Build console output for commit: 84f810bdc0d7872ab44368dcbf783f9050e5c549 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by RyanDowne** [comment]
Does cellscience look a lot different with this change than it does in master?

**Comment by zzmp** [comment]
@RyanDowne no, it looks the same.

I would expect it to look different if it heavily used point or spot lights. @imgntn says there is only one point light, at 3110,13660,3785. I tested at that location and could not notice a difference.

He also mentioned spotlights in demo along the walls, and a point in the fireplace. I looked there, and the falloff of the light looks much more believable with this PR (although we may want to turn up their intensities to make them stand out more). Notice the gradients and the shadow cast by the stove between the point light and the floor.

Current build

PR build

File: libraries/render-utils/src/point_light.slf [hunk 7]
```diff
@@ -48,7 +48,7 @@ void main(void) {
     vec3 fragLightVec = getLightPosition(light) - fragPos.xyz;
 
     // Kill if too far from the light center
-    if (dot(fragLightVec, fragLightVec) > getLightSquareRadius(light)) {
+    if (dot(fragLightVec, fragLightVec) > getLightCutoffRadius(light)) {
```
Comment (samcake) [inline libraries/render-utils/src/point_light.slf]: The test is comparing the square of the fragToLight vector length against the LightMaxRadius
It should be the square of the LightMaxRadius

File: libraries/render-utils/src/spot_light.slf [hunk 8]
```diff
@@ -47,7 +47,7 @@ void main(void) {
     vec3 fragLightVec = getLightPosition(light) - fragPos.xyz;
 
     // Kill if too far from the light center
-    if (dot(fragLightVec, fragLightVec) > getLightSquareRadius(light)) {
+    if (dot(fragLightVec, fragLightVec) > getLightCutoffRadius(light)) {
```
Comment (samcake) [inline libraries/render-utils/src/spot_light.slf]: The test is comparing the square of the fragToLight vector length against the LightMaxRadius
It should be the square of the LightMaxRadius

**Comment by hifi-gustavo** [comment]
Build console output for commit: bab29a0d3e77396f7b232d9a4f4249cf3117fe5d available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c85d0edb1cea1add61dd3cb88f074213fb1b0a66 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7ccb7e1c79221f0824b55b70f4866c21885f0fa9 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by zzmp** [comment]
*Screenshot from 7ccb7e1*
PR build

**Comment by zzmp** [comment]
This reimplements light falloff by scaling/biasing it to a cutoff radius, determined as a function of light source radius (`Lr`) and intensity (`Li`). This avoids the drastic light drop currently in master, but makes the light curve less predictable with high (`Li/Lr^2`) as central regions become drastically more lit.

The maximum radius cutoff (`Rmax`) was originally implemented to bound light without shadowing, and this diverges from that. We're considering making light curves more predictable, so **I am marking this DNM until this is further planned**.

Other things on the table:
- Constant intensity cutoff (`Imax`) instead of a user-defined `Rmax`, so that scaling/biasing is consistent.
- Exposing `Lr` so that a user-defined `Lr` can define the light curve.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d70bc59c851f9c5dbff25dd237335c42dc75421a available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 868e2ec96157f569bb49000e72a31b35672f44b4 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 77ad4039bcde78d5317e9e2837a6ed758adc46af available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dacf9f712361dbfb7f2a512a4207ef05a428b977 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bff4cb07d92a34355e580690a452c0443b492ebc available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bff4cb07d92a34355e580690a452c0443b492ebc available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e0fdf9cb2a1fca2c33f349b18301852ded84b81c available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 78de9cf30ce7d71db86f696fe0cc53d7e721f430 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1fab85e8f5200e119e838ef19cc43d4230eff61b available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fe2de118cae99dacead634f2afe032540540716c available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fe2de118cae99dacead634f2afe032540540716c available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2534bb14f3de6a1936ab0b5f2ef9dcf90d05d47f available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by RyanDowne** [comment]
On PC I don't see the light icons when I have Show Lights in Edit Mode, selected in the edit menu. Select a light, make it a spotlight, zoom out and you will see the cutoff lines but if you zoom in you will see them disappear.

**Comment by zzmp** [comment]
You'll want to use this version of edit.js for testing:
https://rawgit.com/zzmp/hifi/2534bb14f3de6a1936ab0b5f2ef9dcf90d05d47f/examples/edit.js
If the latest commit changes, ou'll need to update this.

**Comment by zzmp** [comment]
@RyanDowne I couldn't repro. I am able to see both the light icon, and the cutoff lines, in both current build and the latest from this PR. Are there more repro steps? Do you have screenshots?

**Comment by hifi-gustavo** [comment]
Build console output for commit: c2ed5e547459003d90ca65f8010469369079e3e3 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by cozza13** [comment]
I have tested this and it works

**Comment by RyanDowne** [comment]
I'd really like to see this merged before the 10:00 today. What else needs testing @chris?

File: libraries/entities/src/EntityPropertyFlags.h [hunk 9]
```diff
@@ -167,6 +167,8 @@ enum EntityPropertyList {
 
     PROP_COLLISION_MASK, // one byte of collision group flags
 
+    PROP_SURFACE_RADIUS, // for Light entity
```
Comment (ZappoMan) [inline libraries/entities/src/EntityPropertyFlags.h]: Why do we need this new property? Can't the radius be calculated by the extent (bounds) of the light object? Do you actually need to know both the total volume (as defined by the bounds) and some additional radius?

**Comment by ZappoMan** [comment]
@RyanDowne - I have a couple questions about this PR... now that I'm looking at it closer, I don't understand why we need to add this new property...

**Comment by ZappoMan** [comment]
After Chatting with @zzmp I now understand the intent of this new property. He's going to rename it to "falloff" to make it more clear, what it does... and then this is good to go in my view.

File: libraries/entities/src/EntityPropertyFlags.h (hunk 9 above)
Comment (ZappoMan) [inline libraries/entities/src/EntityPropertyFlags.h]: Zach will change the name of this to PROP_FALLOFF and adjust the PR accordingly... then we'll be good to go.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 32ce7c6eab7ad6d0a5e72a9a32912f2760a3ded1 available at: https://gist.github.com/111dab1e3c6e73c8675f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7078.dmg

**Comment by zzmp** [comment]
@ZappoMan still looks good on Windows. Default lights are 4x4x4 with falloff of 0.1, and I like how that looks too in my localhost. Not sure how it'll look on demo.

**Comment by zzmp** [comment]
Some Q&A:

[8:45 AM] 
zappoman why can't surface radius be calculated by the extents of the object?
[8:50 AM] 
why can't that information come from the bounds of the light?
what does it mean to have a "surface radius" set for a light that is different than the bounds of the light? Is that concept meaningful to a content creator?
[9:08 AM] 
zach If surface radius were proportional to extents/bounds (correct me if they are different) then lights would behave differently based on size. It would be deterministic, but not intuitive. 
The bound is meant to be the limit of what is lit, as it is currently. 
The surface radius is only called such because of it's physical analogue. It controls the falloff of the light. A larger surface radius means the light will remain more intense longer. 
Really, all of our lights (in this PR) have no physical surface radius, only a corresponding falloff. A more apt name might just be "falloff".
[9:15 AM] 
zappoman ok -- so you're saying that it's meaningful to have both bounds and the additional "falloff" radius?
[9:15 AM] 
so if I have a light with a bounds of 10 meters radius (20m x 20m x 20m bounds) (edited)
[9:16 AM] 
and I set my "surface radius" to say 5m -- what am I expecting the light to look like?
[9:17 AM] 
when I set my "surface radius" to 9m -- what am I expecting the light to look like now?
[9:17 AM] 
when I set my "surface radius" to 15m (larger than the bounds) -- what am I expecting the light to look like now?
[9:18 AM] 
zach Relative to a light with surface radius at 0.1 (default) it will stay much brighter longer
[9:19 AM] 
Because it is treated as if there is a larger light source.
[9:21 AM] 
As I understood it in writing this, having bounds explicitly is our mechanism for limiting the spread of the light so it only needs to be considered locally and won't leak through walls. It's not actually related to the lights behavior. It just acts as a strict bound should the light still be visible at that distance.
[9:21 AM] 
zappoman certainly that statement is true
[9:22 AM] 
in your new model -- what does it mean to have a surface radius larger than the bounds?
[9:22 AM] 
does it mean we get a sharp edge at the bounds?
[9:23 AM] 
that's not bad per se... just wanting to understand it
[9:23 AM] 
zach The other important distinction between this light and reality is that it is biased. If surface radius is 5m, it will falloff from the point as if it are falling off from a surface in reality. There is no light bulb, so there is no maximally lit area inside of the surface radius.
[9:23 AM] 
(Again, why the naming is borky.)
[9:23 AM] 
zappoman you picked the name. :wink: we can always pick a better one. :wink:
[9:24 AM] 
but I'd rather get the right name ... in this PR...
[9:25 AM] 
zach Yes to the sharp edge. It means the light will still be bright at the extent, so it will be a noticeable cutoff.
[9:26 AM] 
zappoman so - maybe we change the name of this property to "falloff" and/or "fallofRadius"

**Comment by AlanZimmerman** [comment]
As far as I can tell, this seems to behave as it's supposed to on OSX. Spawned a spotlight, moved, rotated, scaled.

**Pull request merged**